docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+    value = integrations.mysql.server1.targets
+}
+
+integrations.mysql "server1" {
+    username = argument.username.value
+    password = argument.password.value
+}
+
+```
+
+```river
+// parent
+
+local.file "mysql" {
+    filename = "/test/mysql.river"
+}
+
+module.single "mysql" {
+    content = local.file.mysql.content
+    arguments = {
+        {
+            "password" = PASSWORD,
+            "username" = USERNAME,
+        }
+    }
+}
+
+prometheus.scrape "scraper" {
+    targets = module.single.mysql.exports.targets
+}
+
+```
+
+## Limitations
+
+* Duplicate modules cannot be nested, this may or may not be enforced by the system
+* Singleton components are not supported at this time. Example [node_exporter](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.integration.node_exporter/).
+* Modules will not prevent competing resources, such as starting a server on the same port
+* [Configuration blocks](https://grafana.com/docs/agent/latest/flow/reference/config-blocks/#configuration-blocks) will not be supported. 
+* Arguments and exports within a module must be unique
+
+## Proposal
+
+Add the ability to load `modules` as subgraphs to the primary `graph`. Modules may call other modules within a reasonable stack size depth. Modules are represented as a river string that is interpreted with a defined set of arguments and exports.
+
+The initial component will be `module.single` that will load a single module. Internally these modules will be namespaced so they cannot affect children or parent graphs except via arguments and exports.
+
+Modules will have access to any standard function and any other component exempting singletons. Internally each component in the module will have an `id` that is prepended with the parent's `id` for identification purposes outside of the module. Within the module a component can reference another sibling component normally. There are no known limits on the datatype that a module can use as an argument or export.
+
+`modules.multiple` will use the key of the `map(string)` to uniquely identify a specific module.
+
+### Component Options
+
+
+Given the above example, the `id` of `integrations.mysql "server1"` would be `module.single.mysql.integrations.mysql.server1`. The `data-agent` field would also be prefixed. There are some inherent issues, deeply nested metrics are likely to run into Prometheus label value limits. On Windows platforms there could be issues with the `data-agent` length. These are issues that currently exist in Agent Flow but are more easily hit using deeply nested modules.
+
+
+### Failure Modes
+
+#### Option 1: When a module fails then fail itself and children.
+
+If an error occurs while re-evaluating a module then the module marks itself as unhealthy and unloads the original module and all children.
+
+*Pros*
+
+* Simple to implement
+* Easy to understand
+
+*Cons*
+
+* One failure mode can cascade
+
+#### Option 2: Modules keep last good value
```

Comment:
I'm in favor of this. This is how the Flow main file works today. Failing the module with its children sounds potentially disastrous.

Reply:
This is the ideal situation, and the one we can start with but I think there is some flexibility here. In the current reload we only have to check to see if the config is valid once, in modules that can happen many times and we will need to keep the many old and new sub-graphs running. I have concerns this may lead to duplication.

Reply:
I think if we go with this approach, we would have to indicate to the users which modules/components failed to load  the latest value.

What are the main reasons for not being able to load the latest value? Is it only due to invalid config (e.g. River syntax issues)?

The safest approach imo would be that if a config reload fails a subsection of the graph, the whole graph needs to run on the last valid state, and not just a subsection of it.